inc/zoo/swar/SWAR.h

```diff
@@ -83,99 +88,153 @@ namespace detail {
+        auto invertedMask = filter << (NBits * position);
         auto mask = ~invertedMask;
         return SWAR(m_v & mask);
     }
 
-    constexpr int top() { return msb_index(m_v) / Size; }
-    constexpr int fastIndex() { return __builtin_ctzll(m_v) / Size; }
+    /// Notably, this is the SWAR lane index that contains the MSB.
+    constexpr int top() const noexcept { return msb_index(m_v) / NBits; }
+    /// Notably, this is the SWAR lane index that contains the lsb.
+    constexpr int lsbIndex() const noexcept { return __builtin_ctzll(m_v) / NBits; }
 
-    constexpr SWAR set(int index, int bit) {
-        return SWAR(m_v | (T(1) << (index * Size + bit)));
+    constexpr SWAR set(int index, int bit) const noexcept {
+        return SWAR(m_v | (T(1) << (index * NBits + bit)));
     }
 
-    constexpr operator bool() const { return m_v; }
-protected:
     T m_v;
 };
 
-template<int Size, typename T>
-struct BooleanSWAR: SWAR<Size, T> {
-    constexpr BooleanSWAR(T v): SWAR<Size, T>(v) {}
+/// Defining operator== on base SWAR types is entirely too error prone. Force a verbose invocation.
+template<int NBits, typename T = uint64_t> constexpr auto horizontalEquality(SWAR<NBits, T> left, SWAR<NBits, T> right) {
+  return left.m_v == right.m_v;
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto isolate(T pattern) {
+  return pattern & ((T(1)<<NBits)-1);
+}
+
+template<typename T = uint64_t>
+constexpr auto clearLSB(T v) {
+  return v & (v - 1);
+}
+
+template<typename T = uint64_t>
+constexpr auto isolateLSB(T v) {
+  return v & ~clearLSB(v);
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto maskLowBits() {
+  return (T(1)<<(NBits-1)) | ((T(1)<<(NBits-1))-1);
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto clearLSBits(T v) {
+  return v &(~(maskLowBits<NBits>() << metaLogFloor(isolateLSB<T>(v))));
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto isolateLSBits(T v) {
+  return v &(maskLowBits<NBits>() << metaLogFloor(isolateLSB<T>(v)));
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto broadcast(SWAR<NBits, T> v) {
+  constexpr T Ones = makeBitmask<NBits, T>(SWAR<NBits, T>{1}.value());
+  return SWAR<NBits, T>(T(v) * Ones);
+}
+
+template<int NBits, typename T>
+struct BooleanSWAR: SWAR<NBits, T> {
+    // Booleanness is stored in MSB of a given swar.
+    static constexpr T MSBs = broadcast<NBits, T>(SWAR<NBits, T>(T(1) << (NBits -1))).value();
+    constexpr explicit BooleanSWAR(T v): SWAR<NBits, T>(v) {}
 
-    constexpr BooleanSWAR clear(int bit) {
-        constexpr auto Bit = T(1) << (Size - 1);
-        return this->m_v ^ (Bit << (Size * bit)); }
-    constexpr int best() { return this->top(); }
+    constexpr BooleanSWAR clear(int bit) const noexcept {
+        constexpr auto Bit = T(1) << (NBits - 1);
+        return this->m_v ^ (Bit << (NBits * bit)); }
+    constexpr int best() const noexcept { return this->top(); }
```

Comment:
Hmmm... this name is not useful.  Mind you, I made a lot of mistakes like these, thanks for going over them

Reply:
agreed